app/controllers/auth_controller.rb

```diff
@@ -4,9 +4,16 @@ class AuthController < ApplicationController
   skip_before_filter :verify_authenticity_token, :only => [:access_token]
 
   def verify_logged_in
-    if current_user.nil?
+
+    if session[:omniauth_origin]
       session[:sso_callback_params] = params
-      session[:sso_application] = application
+      session[:sso_application]     = application
+    else
+      session.delete :sso_callback_params
```

Comment:
Is it possible that some request to this controller will be interleaved with the normal login requests? If that happens the callback_params might get deleted.

Reply:
Are you referring to some ajax request to that route? I see in portal-pages the following, though this only calls if it thinks it is logged in:

```
  componentDidMount: function () {
    window.addEventListener('resize', this.handleResize.bind(this));
    if (this.state.loggedIn) {
      var self = this;
      jQuery.ajax({
        url: '/auth/user',  // TODO: replace with Portal.API_V1 constant when available
        dataType: 'json'
      }).done(function (data) {
        self.setState({userId: data.id});
        fadeIn(this, 1000);
      }.bind(this));

```

Aside: I wonder if it is logged in and it needs that id, then maybe it should already be made available without the extra server call.

But either way, would it not have the same session environment once the oauth flow has been initiated?

Reply:
I was thinking both of ajax, and also the case where multiple tabs are open.  As you say, these requests would have the same session environment. So if the an errant call to /user/auth happens after oauth_authorize, but before one of the calls that logs in the user, then the flow will be broken.

I'm not sure I have this right, but I think sso_callback_params are only used when the portal is acting as an authentication provider for something else. Typically that is LARA.

You might have already figured this out, but it helps me reason about it.  The flow for logging in from LARA is:
1. user goes to LARA and chooses login with learn portal
2. this takes them to this url authoring.concord.org/users/auth/cc_portal_learn
3. which redirects them to a URL like: https://learn.concord.org//auth/concord_id/authorize?response_type=code&client_id=authoring&redirect_uri=http%3A%2F%2Fauthoring.concord.org%2Fusers%2Fauth%2Fcc_portal_learn%2Fcallback&state=1234
4. And that is handled by AuthController#oauth_authorize
5. if the user is not logged into learn, then this will save the params in session[:sso_callback_params] and then redirects them to /auth/login
6. AuthController#login checks again to see if the user is logged in. In this flow the user is not logged in. The 'oauth_authorize' call in step 5 already checked that.
7. Since the user isn't logged in, AuthController#login renders a form that will POST their login credentials to: /users/sign_in
8. /users/sign_in processed by the DeviseSessionsController which calls after_sign_in_path_for
9. after_sign_in_path_for looks for these sso_callback_params, if it finds them then it creates the AccessGrant, and redirects back to the redirect stored in the sso_callback_params
10. this redirect should be (taken from url in step 3): authoring.concord.org/users/auth/cc_portal_learn/callback

The most complex flow here is if the user chooses to sign in with google on step 7. But I'm done for the night. I'll try to write up that flow in the morning.

I was going to suggest the following approach, so we can drop the use of session[:sso_callback_params], but then I remembered about the possibility of google login.

---
Half baked idea (won't work because of google auth):

add these params onto the redirect in step 5. Then the form rendered in step 5 can include these params as hidden fields in the form. And then the after_sign_in_path_for method could look for them in params instead of in the session.

But I haven't run through all of the cases, so perhaps this approach won't work. For example if the user is already logged into the portal. I think that will work, but I'm not 100%.
Also, what will happen if the user gets to step 7. Then leaves the page open and forgets about it. Now they open a new tab and login to the portal. Currently they will get redirected to LARA and if they don't have permission they will see a weird error message.  If the params are saved in the /auth/login form, the user will not get redirected to LARA when they login somewhere else.  But what if the user then later finds the /auth/login tab with the login form from step 7 and tries to log in there? That is an edge case, but it should at least not throw an error.

Reply:
Continuing documenting the flow. If a user chooses to sign in with google on Step 7.
The steps below replace all of the steps after 7 above:

8: request to https://learn.concord.org/users/auth/google is sent
9: this does a redirect to google with a URL like: https://accounts.google.com/o/oauth2/auth?access_type=offline&client_id=1234.apps.googleusercontent.com&redirect_uri=https%3A%2F%2Flearn.concord.org%2Fusers%2Fauth%2Fgoogle%2Fcallback&response_type=code&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email
10: the user is lead through the google oauth forms...
11: then the user is redirected back to learn with a url like:
https://learn.concord.org/users/auth/google/callback?code=1234#
12: something happens in the portal that I'm not entirely sure about. I believe whatever happens triggers a call to after_sign_in_path_for.
13: after_sign_in_path_for finds the sso_callback_params 
14: this triggers a redirect back to lara: https://authoring.concord.org/users/auth/cc_portal_learn/callback?code=1234&response_type=code&state=1234

So in order to get rid of storing the sso_callback_params in the session we could:
A. pass them on to google encoded in the redirect_uri that is send to google in step 9. 
B. perhaps google supports a state parameter which it would send back to us with the callback. If so that state could either directly include the parameters, or we could save the parameters in some database entry which is referenced by the state.

---

After going through all of this, I think your approach in this PR is OK.  At least as far as the cases I can think of.  It is possible the user won't get redirected back to LARA. But I think that should be rare. 

I do still want to stop using the session to store the sso_callback_params, but that can be done separately. I want to do that work, not so much to fix the issue of this PR, but to fix the issue of users being confused when they have multiple tabs and they are working with multiple users in lara and the portal. In that situation having session parameters can be confusing and cause unexpected things. Storing the info in the URL or in hidden form fields is less confusing.